File: compilers/layercss-ts/src/compiler.ts
import * as fs from 'fs';
import * as path from 'path';

function compileLYC(inputFile: string, outputFile: string): void {
  try {
    const lycContent = fs.readFileSync(inputFile, 'utf-8');
    const cssContent = processLYC(lycContent);
    fs.writeFileSync(outputFile, cssContent);
    console.log(`Archivo CSS generado: ${outputFile}`);
  } catch (error) {
    console.error(`Error al compilar: ${(error as Error).message}`);
  }
}

function processLYC(lycContent: string): string {
  let cssOutput = '';
  const globalVariables: { [key: string]: string } = {};

  // Eliminar comentarios
  lycContent = lycContent.replace(/\/\/.*|\/\*[\s\S]*?\*\//g, '');

  // Procesar variables globales
  const globalVarMatch = lycContent.match(/^(--[a-zA-Z0-9-]+:\s*[^;]+;)/gm);
  if (globalVarMatch) {
    for (const line of globalVarMatch) {
      const [key, value] = line.split(':').map(part => part.trim());
      globalVariables[key] = value.replace(';', '');
    }
    lycContent = lycContent.replace(/^(--[a-zA-Z0-9-]+:\s*[^;]+;)/gm, '').trim();
  }

  // Procesar bloques
  const blocks = lycContent.split(/(@layer\s+\w+\s*\{|\})/).filter(block => block.trim() !== '');

  for (let i = 0; i < blocks.length; i++) {
    const block = blocks[i].trim();

    if (block.startsWith('@layer')) {
      const layerName = block.match(/@layer\s+(\w+)/)?.[1];
      const layerContent = blocks[++i].trim();
      cssOutput += processBlock(layerContent, globalVariables);
    } else {
      cssOutput += processBlock(block, globalVariables);
    }
  }

  return minifyCSS(cssOutput);
}

function processBlock(blockContent: string, variables: { [key: string]: string }): string {
  // Procesar variables locales
  const localVariables: { [key: string]: string } = {};
  blockContent = blockContent.replace(/(--[a-zA-Z0-9-]+:\s*[^;]+;)/g, (match) => {
    const [key, value] = match.split(':').map(part => part.trim());
    localVariables[key] = value.replace(';', '');
    return '';
  });

  // Combinar variables locales y globales
  const combinedVariables = { ...variables, ...localVariables };

  // Reemplazar variables en el bloque
  let processedBlock = blockContent;
  for (const [varName, varValue] of Object.entries(combinedVariables)) {
    processedBlock = processedBlock.replace(new RegExp(`var\\(${varName}\\)`, 'g'), varValue);
  }

  return processedBlock;
}

function minifyCSS(css: string): string {
  return css.replace(/\s+/g, ' ').trim();
}

// Ejecutar el compilador
const inputFile = path.join(__dirname, '../examples/example1.lyc');
const outputFile = path.join(__dirname, '../examples/example1.css');
compileLYC(inputFile, outputFile);

File: compilers/layercss-python/src/compiler.py
import re
import os

def compile_lyc(input_file, output_file):
    try:
        with open(input_file, 'r', encoding='utf-8') as file:
            lyc_content = file.read()
        css_content = process_lyc(lyc_content)
        with open(output_file, 'w', encoding='utf-8') as file:
            file.write(css_content)
        print(f"Archivo CSS generado: {output_file}")
    except Exception as e:
        print(f"Error al compilar: {e}")

def process_lyc(lyc_content):
    css_output = ""
    global_variables = {}

    # Eliminar comentarios
    lyc_content = re.sub(r'//.*|/\*[\s\S]*?\*/', '', lyc_content)

    # Procesar variables globales
    global_var_matches = re.findall(r'^--([a-zA-Z0-9-]+):\s*([^;]+);', lyc_content, re.MULTILINE)
    for var_name, var_value in global_var_matches:
        global_variables[var_name] = var_value.strip()
    lyc_content = re.sub(r'^--([a-zA-Z0-9-]+):\s*([^;]+);', '', lyc_content, flags=re.MULTILINE).strip()

    # Procesar bloques
    blocks = re.split(r'(@layer\s+\w+\s*\{|})', lyc_content)
    blocks = [block.strip() for block in blocks if block.strip()]

    i = 0
    while i < len(blocks):
        block = blocks[i]
        if block.startswith('@layer'):
            layer_name = re.match(r'@layer\s+(\w+)', block).group(1)
            layer_content = blocks[i + 1]
            css_output += process_block(layer_content, global_variables)
            i += 2
        else:
            css_output += process_block(block, global_variables)
            i += 1

    return minify_css(css_output)

def process_block(block_content, variables):
    local_variables = {}
    block_content = re.sub(r'--([a-zA-Z0-9-]+):\s*([^;]+);', lambda match: local_variables.update({match.group(1): match.group(2).strip()}) or '', block_content)

    # Combinar variables locales y globales
    combined_variables = {**variables, **local_variables}

    # Reemplazar variables
    for var_name, var_value in combined_variables.items():
        block_content = re.sub(rf'var\(--{var_name}\)', var_value, block_content)

    return block_content

def minify_css(css):
    return re.sub(r'\s+', ' ', css).strip()

# Ejecutar el compilador
input_file = os.path.join(os.path.dirname(__file__), '../examples/example1.lyc')
output_file = os.path.join(os.path.dirname(__file__), '../examples/example1.css')
compile_lyc(input_file, output_file)